IdentityCore/src/controllers/MSIDSilentController.m

```diff
@@ -118,33 +118,84 @@ - (void)acquireTokenWithRequest:(MSIDSilentTokenRequest *)request
-            if (!ssoResult && (ssoError.code == MSIDErrorSSOExtensionUnexpectedError))
-            {
-                completionBlock(result, error);
-            }
-            else
+            // If ssoError presented already, skip broker when came back from local RT fallback
+            if (!ssoResult && (self.requestParameters.allowUsingLocalCachedRtWhenSsoExtFailed || ssoError.code == MSIDErrorSSOExtensionUnexpectedError))
             {
-                completionBlock(ssoResult, ssoError);
+                // Skip duplicate local cache lookups
+                request.forceRefresh = YES;
+                request.skipLocalRt = NO;
+                [request executeRequestWithCompletion:^(MSIDTokenResult *localRtResults, NSError *localRtError)
+                {
+#if !EXCLUDE_FROM_MSALCPP
+                    [self addTelemetryEvent:localRtResults isSsoExtFallBack:YES error:localRtError];
+#endif
+                    self.currentRequest = nil;
+                    [self completionHandler:localRtResults
+                                  ssoResult:nil
+                                      error:localRtError
+                                   ssoError:ssoError
+                            completionBlock:completionBlock];
+                }];
+                return;
             }
+            
+            self.currentRequest = nil;
+            [self completionHandler:nil
+                          ssoResult:ssoResult
+                              error:error
+                           ssoError:ssoError
+                    completionBlock:completionBlock];
         };
 
         [self.fallbackController acquireToken:completionBlockWrapper];
     }];
 }
 
+#if !EXCLUDE_FROM_MSALCPP
+- (void)addTelemetryEvent:(MSIDTokenResult *)result
+         isSsoExtFallBack:(BOOL)isSsoExtFallBack
+                    error:(NSError *)error
+{
+    MSIDTelemetryAPIEvent *telemetryEvent = [self telemetryAPIEvent];
+    [telemetryEvent setUserInformation:result.account];
+    [telemetryEvent setIsExtendedLifeTimeToken:result.extendedLifeTimeToken ? MSID_TELEMETRY_VALUE_YES : MSID_TELEMETRY_VALUE_NO];
+    if(isSsoExtFallBack)
+    {
+        [telemetryEvent setSsoExtFallBackFlow:1];
+    }
+    
+    [self stopTelemetryEvent:telemetryEvent error:error];
+}
+#endif
+
+- (void)completionHandler:(nullable MSIDTokenResult *)result
+                ssoResult:(nullable MSIDTokenResult *)ssoResult
+                    error:(nullable NSError *)error
+                 ssoError:(nullable NSError *)ssoError
+          completionBlock:(nonnull MSIDRequestCompletionBlock)completionBlock
+{
+    if (result)
+    {
+        completionBlock(result, error);
+    }
+    else if (!ssoResult && (ssoError.code == MSIDErrorSSOExtensionUnexpectedError))
+    {
+        // We don't have any meaningful information from fallback controller (edge case of SSO error) so we use the local controller result earlier
+        completionBlock(result, error);
```

Comment:
I don't understand this - we seem to return both result and error

Reply:
Updated with more explicitly code